src/JuliusSweetland.OptiKey.Core/UI/Views/Keyboards/Common/DynamicKeyboard.xaml.cs

```diff
@@ -326,13 +331,488 @@ private void SetupErrorLayout(string heading, string content)
+                            openGrid.ElementAt(row).Remove(col);
+                    }
+                }
+                SetupDynamicItem(dynamicItem, minKeyWidth, minKeyHeight);
+            }
+            //end section 1: processing items with a reserved row and column
+
+            //begin section 2: processing items that need a row and/or column assigned
+            //the items are processed in the same order they were listed in the xml file
+            //if an item has a row or column designation it is treated as an indication to jump forward 
+            //to that row or column and mark any/all skipped spaces as empty
+            foreach (XmlDynamicItem dynamicItem in itemPosition.Where(x => !(x.Row > -1 && x.Col > -1)))
+            {
+                var vIndex = itemPosition.IndexOf(dynamicItem);
+                var vLabel = "Suggestion";
+                if (dynamicItem is XmlDynamicKey dynamicKey) { vLabel = dynamicKey.Label; }
+                else if (dynamicItem is XmlDynamicScratchpad) { vLabel = "Scratchpad"; }
+
+            //find first row with enough available width for the item
+            findAvailableRow:
+                //if row is designated then block all preceding columns in all preceding rows
+                if (dynamicItem.Row > -1)
+                {
+                    for (int row = 0; row < dynamicItem.Row; row++)
+                    {
+                        if (openGrid.ElementAt(row).Count > 0)
+                            openGrid.ElementAt(row).RemoveRange(0, openGrid.ElementAt(row).Count);
+                    }
+                }
+                var vItemColumn = 0;
+                var vRowsConfirmed = 0;
+                var startRow = openGrid.FindIndex(x => (x.Count() >= dynamicItem.Width));
+                if (startRow < 0 || (dynamicItem.Row > -1 && startRow > dynamicItem.Row)) //if not found
+                {
+                    SetupErrorLayout("Invalid keyboard file", "Insufficient space to position item "
+                        + (vIndex + 1) + " of " + itemPosition
+                        + " labeled '" + vLabel + "' having width "
+                        + dynamicItem.Width + " and height " + dynamicItem.Height);
+                    return false;
+                }
+                for (int row = startRow; row < keyboard.Grid.Rows; row++)
+                {
+                    if (dynamicItem.Col > -1)
+                    {
+                        //if column exists then block all preceding columns in the row
+                        if (openGrid.ElementAt(row).Exists(x => x == dynamicItem.Col))
+                            openGrid.ElementAt(row).RemoveAll(x => x < dynamicItem.Col);
+                        //else block the whole row and check another 
+                        else
+                        {
+                            openGrid.ElementAt(row).RemoveRange(0, openGrid.ElementAt(row).Count);
+                            goto findAvailableRow;
```

Comment:
Although there's technically nothing wrong with a goto statement, flow control using goto statements is not really correct. Would it be possible to refactor this logic into a method instead?

Reply:
LOL! You purist. No infinite loops. No goto statements.
You want a refactor? I'll give you a refactor.